[2기-C 김승은] Spring Boot JPA 게시판 구현 코드 리뷰 부탁드립니다

julie0005 opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- SpringBoot Part4 SpringData JPA 게시판 Rest API 만들기 과제입니다.
## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
### **SpringDataJPA 를 설정한다.**

- datasource : h2 or mysql

### **엔티티를 구성한다**

- 회원(User)
    - id (PK) (auto increment)
    - name
    - age
    - hobby
    - **created_at**
    - **created_by**
- 게시글(Post)
    - id (PK) (auto increment)
    - title
    - content
    - **created_at**
    - **created_by**
- 회원과 게시글에 대한 연관관계를 설정한다.
    - 회원과 게시글은 1:N 관계이다.
- 게시글 Repository를 구현한다. (PostRepository)

### **API를 구현한다.**

- 게시글 조회
    - 페이징 조회 (GET "/posts")
    - 단건 조회 (GET "/posts/{id}")
- 게시글 작성 (POST "/posts")
- 게시글 수정 (POST "/posts/{id}")

### **REST-DOCS를 이용해서 문서화한다.**
## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
### 궁금한 점
1. 이전 과제에서는 repository test를 별도로 진행했었기에 mockito로 사용하여 값이 나온다는 것을 전제로 테스트를 진행했었습니다. 이번 과제에서는 jpa를 사용하기 때문에 mockito 사용 없이 직접 service의 메소드를 호출하여 repository에서 올바른 값을 보내는지 확인했습니다. jpa를 사용할 때 mockito를 사용하면 repository를 따로 해주는건가요? service 테스트의 목적에만 맞으면 mockito 없이 테스트 해도 상관 없나요?

2. entity validation은 비즈니스 로직이 커질 것을 생각하여 validate 함수를 private으로 가지게 하였고, dto에서 validation은 annotation을 활용하여 간단한 것만 진행했습니다. 둘다 annotation을 사용하는게 맞는걸까요?

3. Rest-docs가 swagger나 수기로 만든 것들에 비해 상당히 가독성이 좋지 않은데 실무에서 실제로 프론트분들이 rest-docs를 보고 작성하시나요?

4. 미리 만들어둔 Response 포맷을 ResponseEntity로 한번 더 감쌌습니다. Response 내의 status 값을 설정하더라도 실제로 http 헤더의 status를 조작하기 위해서는 ResponseEntity가 필요했습니다.  실무에서도 ResponseEntity로 미리 만들어둔 format을 한번더 감싸나요?

julie0005 committed:
init : init project

julie0005 committed:
feat : 연관관계 맵핑

julie0005 committed:
feat : BaseEntity Auditing 수정, 초기 데이터 삽입

julie0005 committed:
feat : PostService 작성,수정,조회 구현

julie0005 committed:
fix : 게시글 작성자 값은 수정하지 못하도록 바꿈

julie0005 committed:
test : PostServiceTest 구현

julie0005 committed:
feat : PostController 구현

julie0005 committed:
test : PostController test 구현, update api 수정

julie0005 committed:
doc : api 명세 작성

julie0005 committed:
refactor : Post Entity Validation

julie0005 committed:
refactor : Dto Validation, Controller 실패 test 추가

park0503 commented on src/main/java/com/prgrms/boardjpa/Response.java:
```diff
@@ -0,0 +1,32 @@
+public class Response<T> {
+    private int statusCode;
+    private T data;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
+    private LocalDateTime serverDatetime;
```
ResponseBody에서는 따로 `@JsonFormat`을 지정해주지 않아도 값이 잘 전송이 되더라구요!

park0503 commented on src/test/java/com/prgrms/boardjpa/post/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,127 @@
+
+    }
+
+    @AfterEach
+    void tearDown() {
+        userRepository.deleteAll();
+        postRepository.deleteAll();
+    }
```
테스트 클래스에 `@Transactional` 어노테이션을 달아주면 각 테스트들이 별개의 Transaction으로 실행되고 각 테스트가 끝날 때마다 자동으로 Rollback이 되기에 다음과 같이 직접 Table을 비워주는 구문을 쓰지 않아도 되게 됩니다. 참고하시면 좋을 것 같습니다!

julie0005 replied:
테스트에서 transaction을 활용할 생각을 못해보았네요! 감사합니다

julie0005 replied:
test에서 transactional을 사용할 때의 [유의점](https://javabom.tistory.com/103)입니다! 참고해주시면 좋을 것 같습니다!

park0503 commented on src/main/java/com/prgrms/boardjpa/domain/Post.java:
```diff
@@ -42,8 +50,22 @@ private static boolean validateContent(String content){
 
-    public void setAuthor(User user){
-        if(Objects.nonNull(this.author)){
+    public Post() {
+    }
```
기본 생성자의 경우 JPA의 자동 매핑으로 인한 필요로 사용하는 건데, 접근 제어자가 public이면 예상치 못하게 사용되어 에러를 유발할 수도 있기 때문에 이를 변경할 필요가 있을 것 같습니다.
따라서 접근 제어자를 protected나 private으로 변경해야 하는데 이에는 [다음 링크](https://erjuer.tistory.com/106) 참고하시면 도움 되실 것 같습니다!

julie0005 replied:
링크 읽어보겠습니다!

park0503 commented on src/main/java/com/prgrms/boardjpa/domain/Post.java:
```diff
@@ -0,0 +1,38 @@
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @JoinColumn(name = "user_id", referencedColumnName = "user_id")
+    private User author;
```
Post에서 User에 CascaseType.ALL을 걸어주는 이유에는 어떤게 있을까요?

julie0005 replied:
이 부분은 실수인 것 같네요ㅠㅠ

park0503 commented on src/main/java/com/prgrms/boardjpa/post/controller/PostController.java:
```diff
@@ -0,0 +1,44 @@
+package com.prgrms.boardjpa.post.controller;
+
+import com.prgrms.boardjpa.Response;
+import com.prgrms.boardjpa.post.dto.PostReqDto;
+import com.prgrms.boardjpa.post.dto.PostResDto;
+import com.prgrms.boardjpa.post.dto.PostUpdateDto;
+import com.prgrms.boardjpa.post.service.PostService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequiredArgsConstructor // 생성자 주입
+public class PostController {
+
+    private final PostService postService;
+
+    @ExceptionHandler
+    private Response<String> handleException(Exception exception) {
+        return Response.fail(HttpStatus.INTERNAL_SERVER_ERROR.value(), exception.getMessage());
+    }
```
`ResponseEntity`를 사용하지 않고 Http Status code를 커스터마이징 할 수 있는 방법으로 [다음과 같은 방법들](https://www.amitph.com/spring-return-specific-http-status/)을 참고하시면 좋을 것 같습니다.
먼저 `@ResponseStatus` 어노테이션을 이용할 수 있는데, 얘는 Controller, Exception Handler, Custom Exception 셋 중 사용 용도에 맞춰 어디에든 붙여 쓸 수 있습니다.
그리고 `ResponseStatusException` 객체를 사용하는 방법도 있더라구요.

julie0005 replied:
감사합니다 참고하겠습니다!

park0503 reviewed (commented):
수고하셨습니다 승은님! 코드리뷰 늦어져서 죄송합니다! ㅜㅜ
전체적으로 코드들이 되게 깔끔하단 느낌을 받았고 디테일한 부분 (Column 최대 길이 지정이나 입력값 최대길이 지정, Validation 등)에 신경을 많이 쓰셨단 것이 눈에 잘 보입니다.
몇몇부분에 코멘트 짧게 남겨놨으니 확인해주시면 감사하겠습니다!

julie0005 committed:
refactor : 상혁님 리뷰 반영
1. NoArgsConstructor protected로 변경 후 생성 코드 수정
2. service에 transactional annotation 추가
3. post -> user cascade 옵션 오류 수정
